File: p2/code/demosaicImage.py
# This code is part of:
#
#   CMPSCI 670: Computer Vision, Fall 2018
#   University of Massachusetts, Amherst
#   Instructor: Subhransu Maji
#
#   Homework 1 

import numpy as np

def demosaicImage(image, method='nn'):
    ''' Demosaics image.

    Args:
        img: np.array of size NxM.
        method: demosaicing method (baseline or nn).

    Returns:
        Color image of size NxMx3 computed using method.
    '''
    # print(image.shape, (image[0]))

    if method.lower() == "baseline":
        return demosaicBaseline(image.copy())
    elif method.lower() == 'nn':
        return demosaicNN(image.copy()) # Implement this
    elif method.lower() == 'linear':
        return demosaicLinear(image.copy()) # Implement this
    elif method.lower() == 'adagrad':
        return demosaicAdagrad(image.copy()) # Implement this
    else:
        raise ValueError("method {} unkown.".format(method))


def demosaicBaseline(img):
    '''Baseline demosaicing.
    
    Replaces missing values with the mean of each color channel.
    
    Args:
        img: np.array of size NxM.

    Returns:
        Color image of sieze NxMx3 demosaiced using the baseline 
        algorithm.
    '''
    mos_img = np.tile(img[:, :, np.newaxis], [1, 1, 3])
    image_height, image_width = img.shape

    red_values = img[0:image_height:2, 0:image_width:2]
    mean_value = red_values.mean()
    mos_img[:, :, 0] = mean_value
    mos_img[0:image_height:2, 0:image_width:2, 0] = img[0:image_height:2, 0:image_width:2]

    blue_values = img[1:image_height:2, 1:image_width:2]
    mean_value = blue_values.mean()
    mos_img[:, :, 2] = mean_value
    mos_img[1:image_height:2, 1:image_width:2, 2] = img[1:image_height:2, 1:image_width:2]

    mask = np.ones((image_height, image_width))
    mask[0:image_height:2, 0:image_width:2] = -1
    mask[1:image_height:2, 1:image_width:2] = -1
    green_values = mos_img[mask > 0]
    mean_value = green_values.mean()

    green_channel = img
    green_channel[mask < 0] = mean_value
    mos_img[:, :, 1] = green_channel

    return mos_img


def demosaicNN(img):
    '''Nearest neighbor demosaicing.
    
    Args:
        img: np.array of size NxM.
    '''

    final_img = np.zeros((img.shape[0], img.shape[1], 3), dtype = float)
    # print(final_img.shape)
    # img = np.expand_dims(img, axis=2)
    print(img.shape)

    for i in range(img.shape[0]):
        for j in range(img.shape[1]):
            if(i%2 == 0 and j%2==0):
                try:
                    final_img[i,j,1] = float((img[i,j+1] +img[i+1, j])/2)
                    final_img[i, j, 0] = img[i+1, j + 1]
                    final_img[i,j,2] = img[i,j]
                except:
                    final_img[i, j, 1] = float((img[i, j - 1] + img[i - 1, j]) / 2)
                    final_img[i, j, 0] = float((img[i - 1, j - 1]))
                    final_img[i, j, 2] = img[i, j]

            elif (i % 2 == 1 and j % 2 == 1):
                try:
                    final_img[i, j, 1] = float((img[i, j + 1] + img[i + 1, j]) / 2)
                    final_img[i, j, 2] = img[i + 1, j + 1]
                    final_img[i, j, 0] = img[i, j]
                except:
                    final_img[i, j, 1] = float((img[i, j - 1] + img[i - 1, j]) / 2)
                    final_img[i, j, 2] = float((img[i - 1, j - 1]))
                    final_img[i, j, 0] = img[i, j]
    print((final_img))
    exit(0)
    return 0


def demosaicLinear(img):
    '''Nearest neighbor demosaicing.
    
    Args:
        img: np.array of size NxM.
    '''
    return demosaicBaseline(img)


def demosaicAdagrad(img):
    '''Nearest neighbor demosaicing.
    
    Args:
        img: np.array of size NxM.
    '''
    return demosaicBaseline(img)

File: p2/code/demosaicImage.m
function output = demosaicImage(im, method)
switch lower(method)
    case 'baseline'
        output = demosaicBaseline(im);
    case 'nn'
        output = demosaicNN(im);         % Implement this
    case 'linear'
        output = demosaicLinear(im);     % Implement this
    case 'adagrad'
        output = demosaicAdagrad(im);    % Implement this
    case 'divide'
    	output = demosaicTransformDivide(im);
    case 'log'
        output = demosaicTransformLog(im);
end

%--------------------------------------------------------------------------
%                          Baseline demosacing algorithm. 
%                          The algorithm replaces missing values with the
%                          mean of each color channel.
%--------------------------------------------------------------------------
function mosim = demosaicBaseline(im)
mosim = repmat(im, [1 1 3]); % Create an image by stacking the input
[h, w] = size(im);

% Red channel (odd rows and columns);
r = im(1:2:h, 1:2:w);
meanValue = mean(mean(r));
mosim(:,:,1) = meanValue;
mosim(1:2:h, 1:2:w,1) = im(1:2:h, 1:2:w);

% Blue channel (even rows and colums);
b = im(2:2:h, 2:2:w);
meanValue = mean(mean(b));
mosim(:,:,3) = meanValue;
mosim(2:2:h, 2:2:w,3) = im(2:2:h, 2:2:w);

% Green channel (remaining places)
% We will first create a mask for the green pixels (+1 green, -1 not green)
mask = ones(h, w);
mask(1:2:h, 1:2:w) = -1;
mask(2:2:h, 2:2:w) = -1;
g = mosim(mask > 0);
meanValue = mean(g);
% For the green pixels we copy the value
greenChannel = im;
greenChannel(mask < 0) = meanValue;
mosim(:,:,2) = greenChannel;

%--------------------------------------------------------------------------
%                           Nearest neighbour algorithm
%--------------------------------------------------------------------------
function mosim = demosaicNN(im)
  mosim = repmat(im, [1 1 3]);
  [h, w] = size(im);
  
  %r channel
  for row = 1:h
      for col =1:w
          if(rem(row,2)==1 && rem(col,2) == 1)
              continue;
          elseif(rem(row,2)==0 && rem(col,2)==1)
              mosim(row,col,1) = mosim(row-1,col,1);
          elseif(rem(row,2)==1 && rem(col,2)==0)
              mosim(row,col,1) = mosim(row,col-1,1);
          elseif(rem(row,2)==0 && rem(col,2)==0)
                  mosim(row,col,1) = mosim(row-1,col-1,1);
          end
      end
  end
  
  %g channel
  for row = 1:h
      for col =1:w
          if((rem(row,2)==0 && rem(col,2)==1) || (rem(row,2)==1 && rem(col,2)==0))
              continue;
          elseif(rem(row,2)==1 && rem(col,2)==1)
              if(col<w)
                  mosim(row,col,2) = mosim(row,col+1,2);
              elseif(col==h)
                  if(row<h)
                      mosim(row,col,2) = mosim(row+1,col,2);
                  else
                      mosim(row,col,2) = mosim(row,col-1,2);
                  end
              end
          elseif(rem(row,2)==0 && rem(col,2)==0)
              mosim(row,col,2) = mosim(row,col-1,2);
          end
          
              
      end
  end  
   %b channel
    for row = 1:h
      for col =1:w
          if(rem(row,2)==0 && rem(col,2)==0)
              continue;%blue pixels
          elseif(rem(row,2)==1 && rem(col,2)==1)
              if(row<h && col<w)
                  mosim(row,col,3) = mosim(row+1,col+1,3);
              elseif(row==h && col~=w)
                  mosim(row,col,3) = mosim(row-1,col+1,3);
              elseif(col==w && row~=h)
                  mosim(row,col,3) = mosim(row+1,col-1,3);
              elseif(row==h && col==w)
                  mosim(row,col,3) = mosim(row-1,col-1,3);
              end
          elseif(rem(row,2)==0 && rem(col,2)==1)
              if(col<w)
                  mosim(row,col,3) = mosim(row,col+1,3);
              elseif(col==w)
                  mosim(row,col,3) = mosim(row,col-1,3);
              end
          elseif(rem(row,2)==1 && rem(col,2)==0)
              if(row<h)
                  mosim(row,col,3) = mosim(row+1,col,3);
              elseif(row==h)
                  mosim(row,col,3) = mosim(row-1,col,3);
              end
          end
      end
    end
%--------------------------------------------------------------------------
%                           Linear interpolation
%--------------------------------------------------------------------------
function mosim = demosaicLinear(im)
mosim = repmat(im, [1 1 3]); % Create an image by stacking the input
[r, g, b] = getRGBValues(im);

% Calculate blue pixels at the red location
blueRed = imfilter(b,[1 0 1; 0 0 0; 1 0 1]/4);
% Calculate blue at green location
blueGreen = imfilter(b+blueRed,[0 1 0; 1 0 1; 0 1 0]/4);
mosim(:,:,3) = b + blueRed + blueGreen;

%Calculate red pixels at the blue location
redBlue = imfilter(r,[1 0 1; 0 0 0; 1 0 1]/4);
%Calculate red pixels at the green locations   
redGreen = imfilter(r+redBlue,[0 1 0; 1 0 1; 0 1 0]/4);
mosim(:,:,1) = r + redBlue + redGreen;

% filter for green is same for red and blue pixels
mosim(:,:,2) = g + imfilter(g, [0 1 0; 1 0 1; 0 1 0]/4);

%--------------------------------------------------------------------------
%                           Adaptive gradient
%--------------------------------------------------------------------------
function mosim = demosaicAdagrad(im)
mosim = repmat(im, [1 1 3]); % Create an image by stacking the input
	[h, w] = size(im);
	[r, g, b] = getRGBValues(im);

	%Calculate the missing blue pixels at the red location
	blueRed = imfilter(b,[1 0 1; 0 0 0; 1 0 1]/4);
	% Calculate blue at green pixels
	blueGreen = imfilter(b+blueRed,[0 1 0; 1 0 1; 0 1 0]/4);
	mosim(:,:,3) = b + blueRed + blueGreen;

	%  calculate the missing red pixels at the blue location
	redBlue = imfilter(r,[1 0 1; 0 0 0; 1 0 1]/4);
	% Calculate the missing red pixels at the green locations   
	redGreen = imfilter(r+redBlue,[0 1 0; 1 0 1; 0 1 0]/4);
	mosim(:,:,1) = r + redBlue + redGreen;
    outputGreen = g;
    
	for row = 1:h
		for column = 1:w
			if(g(row, column) ~= 0)
				continue;
			end
			if(row > 1 && column > 1 && row < h && column < w)
				if( abs(g(row-1, column) - g(row+1, column)) > abs(g(row, column-1) - g(row, column+1)) )
					outputGreen(row, column) = ( g(row, column-1) + g(row, column+1) )/2;
				else
					outputGreen(row, column) = ( g(row-1, column) + g(row+1, column) )/2;
				end
			elseif( row == 1 || row == h)
				if(row == 1)
					rowToBeUsed = row+1;
				else
					rowToBeUsed = row-1;
				end
				if(column == 1)
					outputGreen(row, column) = ( g(rowToBeUsed, column) + g(row, column+1) )/2;
				elseif(column == w)
					outputGreen(row, column) = ( g(rowToBeUsed, column) + g(row, column-1) )/2;
				else
					outputGreen(row, column) =  ( g(rowToBeUsed, column) + g(row, column-1) + g(row, column+1))/3;
				end
			elseif( column == 1 || column == w )
				if(column == 1)
					colToBeUsed = column+1;
				else
					colToBeUsed = column-1;
				end
				outputGreen(row, column) = ( g(row, colToBeUsed) + g(row-1, column) + g(row+1, column))/3;
			end
		end
	end
	mosim(:,:,2) = outputGreen;





function mosim = demosaicTransformDivide(im)
	mosim = repmat(im, [1 1 3]); % Create an image by stacking the input
	[r, g, b] = getRGBValues(im);
	% the filter for green is same for red and blue pixels
	mosim(:,:,2) = g + imfilter(g, [0 1 0; 1 0 1; 0 1 0]/4);
    g = mosim(:,:,2);
    g(~g) = 0.01; %set 0.01 to zero vals to avoid zero division

    r = r ./ g;
    b = b ./ g;

	%Calculate the missing blue pixels at the red location
	blueRed = imfilter(b,[1 0 1; 0 0 0; 1 0 1]/4);
	% Calculate blue at green, by averaging over 4 surrounding blue values
	blueGreen = imfilter(b+blueRed,[0 1 0; 1 0 1; 0 1 0]/4);
	b = b + blueRed + blueGreen;


	%Calculate the missing red pixels at the blue location
	redBlue = imfilter(r,[1 0 1; 0 0 0; 1 0 1]/4);
	% Calculate the missing red pixels at the green locations
	redGreen = imfilter(r+redBlue,[0 1 0; 1 0 1; 0 1 0]/4);
	r = r + redBlue + redGreen;
    mosim(:,:,1) = r .* g; %Transforming back the pixels
    mosim(:,:,3) = b .* g;

function mosim = demosaicTransformLog(im)
	mosim = repmat(im, [1 1 3]); % Create an image by stacking the input
	[r, g, b] = getRGBValues(im);
	% the filter for green is same for red and blue pixels
	mosim(:,:,2) = g + imfilter(g, [0 1 0; 1 0 1; 0 1 0]/4);
    g = mosim(:,:,2);

    g(~g) = 0.01; %set 0.01 to zero vals to avoid zero division

    r = r ./ g;
    r(~r) = 0.1; %set 0.01 to zero vals
    b = b ./ g;
    b(~b) = 0.1;
    r = log(r);
    b = log(b);

	%Calculate the missing blue pixels at the red location
	blueRed = imfilter(b,[1 0 1; 0 0 0; 1 0 1]/4);
	% Calculate blue at green, by averaging over 4 surrounding blue values
	blueGreen = imfilter(b+blueRed,[0 1 0; 1 0 1; 0 1 0]/4);
	b = b + blueRed + blueGreen;


	%Calculate the missing red pixels at the blue location
	redBlue = imfilter(r,[1 0 1; 0 0 0; 1 0 1]/4);
	% Calculate the missing red pixels at the green locations
	redGreen = imfilter(r+redBlue,[0 1 0; 1 0 1; 0 1 0]/4);
	r = r + redBlue + redGreen;
    %Transforming back
    mosim(:,:,1) = exp(r) .* g;
    mosim(:,:,3) = exp(b) .* g;

function [r, g, b] = getRGBValues(im)
	[h, w] = size(im);
	r = zeros(h, w);
	r(1:2:h, 1:2:w) = im(1:2:h, 1:2:w);

	b = zeros(h, w);
	b(2:2:h, 2:2:w) = im(2:2:h, 2:2:w);

	g = im;
	g(2:2:h, 2:2:w) = 0;
	g(1:2:h, 1:2:w) = 0;